velox/functions/prestosql/GreatestLeast.cpp

```diff
@@ -0,0 +1,138 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+
+namespace facebook::velox::functions {
+
+namespace {
+
+template <bool>
+class ExtremeValueFunction;
+
+using LeastFunction = ExtremeValueFunction<true>;
+using GreatestFunction = ExtremeValueFunction<false>;
+
+/**
+ * This class implements two functions:
+ *
+ * greatest(value1, value2, ..., valueN) → [same as input]
+ * Returns the largest of the provided values.
+ *
+ * least(value1, value2, ..., valueN) → [same as input]
+ * Returns the smallest of the provided values.
+ **/
+template <bool isLeast>
+class ExtremeValueFunction : public exec::VectorFunction {
+ private:
+  template <typename T>
+  bool shouldOverride(const T& currentValue, const T& candidateValue) const {
+    return isLeast ? candidateValue < currentValue
+                   : candidateValue > currentValue;
+  }
+
+  template <typename T>
+  void applyTyped(
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const {
+    BaseVector::ensureWritable(rows, outputType, context->pool(), result);
+    BufferPtr resultValues =
+        (*result)->as<FlatVector<T>>()->mutableValues(rows.end());
+    T* __restrict rawResult = resultValues->asMutable<T>();
+
+    exec::DecodedArgs decodedArgs(rows, args, context);
+
+    rows.applyToSelected([&](int row) {
+      T currentValue = decodedArgs.at(0)->valueAt<T>(row);
+      for (auto i = 1; i < args.size(); ++i) {
+        auto candidateValue = decodedArgs.at(i)->template valueAt<T>(row);
+        if (shouldOverride(currentValue, candidateValue)) {
+          currentValue = candidateValue;
+        }
+      }
+      rawResult[row] = currentValue;
+    });
+  }
+
+ public:
+  bool isDefaultNullBehavior() const override {
+    return true;
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    switch (outputType.get()->kind()) {
+      case TypeKind::BIGINT:
+        applyTyped<TypeTraits<TypeKind::BIGINT>::NativeType>(
+            rows, args, outputType, context, result);
+        return;
+      case TypeKind::DOUBLE:
+        applyTyped<TypeTraits<TypeKind::DOUBLE>::NativeType>(
+            rows, args, outputType, context, result);
+        return;
+      case TypeKind::VARCHAR:
+        applyTyped<TypeTraits<TypeKind::VARCHAR>::NativeType>(
+            rows, args, outputType, context, result);
+        return;
+      case TypeKind::TIMESTAMP:
+        applyTyped<TypeTraits<TypeKind::TIMESTAMP>::NativeType>(
+            rows, args, outputType, context, result);
+        return;
+      default:
+        VELOX_CHECK(
```

Comment:
VELOX_FAIL("Unsupported input type for {}: {}", isLeast ? "Least" : "Greatest", outputType->toString());